File: frontend/src/app/core/services/admin-artist.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Artist } from '../models/music.model';
import {
  ApiResponse,
  PageResponse,
  PaginationParams,
} from '../models/api.model';
import { environment } from '../../../environments/environment';

export interface ArtistCreateRequest {
  name: string;
  bio?: string;
  avatarUrl?: string;
  isActive?: boolean;
}

export interface ArtistUpdateRequest {
  name: string;
  bio?: string;
  avatarUrl?: string;
  isActive?: boolean;
}

export interface ArtistFilters {
  search?: string;
  isActive?: boolean;
}

export interface ArtistStats {
  artist: Artist;
  totalMusic: number;
  activeMusicCount: number;
}

@Injectable({
  providedIn: 'root',
})
export class AdminArtistService {
  private readonly API_URL = `${environment.apiUrl}/admin/artists`;

  constructor(private http: HttpClient) {}

  getAllArtists(
    searchTerm: string = '',
    isActive?: boolean,
    page: number = 1,
    size: number = 10,
    sort: string = 'name',
    direction: string = 'asc'
  ): Observable<ApiResponse<PageResponse<Artist>>> {
    let httpParams = new HttpParams()
      .set('page', (page - 1).toString())
      .set('size', size.toString())
      .set('sortBy', sort)
      .set('sortDir', direction);

    if (searchTerm.trim()) {
      httpParams = httpParams.set('search', searchTerm.trim());
    }

    if (isActive !== undefined) {
      httpParams = httpParams.set('isActive', isActive.toString());
    }

    return this.http.get<ApiResponse<PageResponse<Artist>>>(this.API_URL, {
      params: httpParams,
    });
  }

  getArtistById(id: number): Observable<ApiResponse<Artist>> {
    return this.http.get<ApiResponse<Artist>>(`${this.API_URL}/${id}`);
  }

  createArtist(request: ArtistCreateRequest): Observable<ApiResponse<Artist>> {
    return this.http.post<ApiResponse<Artist>>(this.API_URL, request);
  }

  updateArtist(
    id: number,
    request: ArtistUpdateRequest
  ): Observable<ApiResponse<Artist>> {
    return this.http.put<ApiResponse<Artist>>(`${this.API_URL}/${id}`, request);
  }

  deleteArtist(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.API_URL}/${id}`);
  }

  toggleActiveStatus(id: number): Observable<ApiResponse<Artist>> {
    return this.http.patch<ApiResponse<Artist>>(
      `${this.API_URL}/${id}/toggle-active`,
      {}
    );
  }

  getArtistStats(id: number): Observable<ApiResponse<ArtistStats>> {
    return this.http.get<ApiResponse<ArtistStats>>(
      `${this.API_URL}/${id}/stats`
    );
  }

  bulkDeleteArtists(artistIds: number[]): Observable<ApiResponse<void>> {
    return this.http.post<ApiResponse<void>>(
      `${this.API_URL}/bulk-delete`,
      artistIds
    );
  }

  bulkToggleActive(
    artistIds: number[],
    active: boolean
  ): Observable<ApiResponse<void>> {
    return this.http.post<ApiResponse<void>>(
      `${this.API_URL}/bulk-toggle-active?active=${active}`,
      artistIds
    );
  }

  // Convenience method to get active artists for dropdowns
  getActiveArtists(): Observable<ApiResponse<PageResponse<Artist>>> {
    return this.getAllArtists(
      '', // searchTerm
      true, // isActive
      1, // page
      1000, // size
      'name', // sort
      'asc' // direction
    );
  }
}

File: backend/src/main/java/com/coconutmusic/service/AdminArtistService.java
package com.coconutmusic.service;

import com.coconutmusic.dto.request.ArtistCreateRequest;
import com.coconutmusic.dto.request.ArtistUpdateRequest;
import com.coconutmusic.entity.Artist;
import com.coconutmusic.exception.BadRequestException;
import com.coconutmusic.exception.ResourceNotFoundException;
import com.coconutmusic.repository.ArtistRepository;
import com.coconutmusic.repository.MusicRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import jakarta.persistence.criteria.Predicate;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@Transactional
public class AdminArtistService {

    @Autowired
    private ArtistRepository artistRepository;

    @Autowired
    private MusicRepository musicRepository;

    /**
     * Get all artists with filtering and pagination
     */
    public Page<Artist> getAllArtists(Pageable pageable, String search, Boolean isActive) {
        Specification<Artist> spec = (root, query, criteriaBuilder) -> {
            List<Predicate> predicates = new ArrayList<>();

            // Search by name
            if (search != null && !search.trim().isEmpty()) {
                String searchPattern = "%" + search.toLowerCase() + "%";
                predicates.add(criteriaBuilder.like(
                    criteriaBuilder.lower(root.get("name")), searchPattern));
            }

            // Filter by active status
            if (isActive != null) {
                predicates.add(criteriaBuilder.equal(root.get("isActive"), isActive));
            }

            return criteriaBuilder.and(predicates.toArray(new Predicate[0]));
        };

        return artistRepository.findAll(spec, pageable);
    }

    /**
     * Get artist by ID
     */
    public Artist getArtistById(Long id) {
        return artistRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Artist", "id", id));
    }

    /**
     * Create new artist
     */
    public Artist createArtist(ArtistCreateRequest request) {
        // Check if artist name already exists
        if (artistRepository.existsByName(request.getName())) {
            throw new BadRequestException("Artist name already exists: " + request.getName());
        }

        Artist artist = new Artist();
        artist.setName(request.getName());
        artist.setBio(request.getBio());
        artist.setAvatarUrl(request.getAvatarUrl());
        artist.setIsActive(request.getIsActive());

        return artistRepository.save(artist);
    }

    /**
     * Update artist
     */
    public Artist updateArtist(Long id, ArtistUpdateRequest request) {
        Artist artist = getArtistById(id);

        // Check if new name already exists (excluding current artist)
        if (!artist.getName().equals(request.getName()) &&
            artistRepository.existsByName(request.getName())) {
            throw new BadRequestException("Artist name already exists: " + request.getName());
        }

        artist.setName(request.getName());
        artist.setBio(request.getBio());
        artist.setAvatarUrl(request.getAvatarUrl());
        artist.setIsActive(request.getIsActive());

        return artistRepository.save(artist);
    }

    /**
     * Delete artist
     */
    public void deleteArtist(Long id) {
        Artist artist = getArtistById(id);

        // Check if artist has music associated
        long musicCount = musicRepository.countByArtistId(id);
        if (musicCount > 0) {
            throw new BadRequestException("Cannot delete artist with associated music. Please reassign or delete the music first.");
        }

        artistRepository.delete(artist);
    }

    /**
     * Toggle active status
     */
    public Artist toggleActiveStatus(Long id) {
        Artist artist = getArtistById(id);
        artist.setIsActive(!artist.getIsActive());
        return artistRepository.save(artist);
    }

    /**
     * Get artist statistics
     */
    public Map<String, Object> getArtistStats(Long id) {
        Artist artist = getArtistById(id);

        Map<String, Object> stats = new HashMap<>();
        stats.put("artist", artist);
        stats.put("totalMusic", musicRepository.countByArtistId(id));
        stats.put("activeMusicCount", musicRepository.countByArtistIdAndIsActiveTrue(id));

        return stats;
    }

    /**
     * Bulk delete artists
     */
    public void bulkDeleteArtists(Long[] artistIds) {
        List<Long> ids = Arrays.asList(artistIds);

        // Check if any artist has associated music
        for (Long id : ids) {
            long musicCount = musicRepository.countByArtistId(id);
            if (musicCount > 0) {
                Artist artist = getArtistById(id);
                throw new BadRequestException("Cannot delete artist '" + artist.getName() + "' with associated music. Please reassign or delete the music first.");
            }
        }

        List<Artist> artists = artistRepository.findAllById(ids);
        artistRepository.deleteAll(artists);
    }

    /**
     * Bulk toggle active status
     */
    public void bulkToggleActive(Long[] artistIds, boolean active) {
        List<Long> ids = Arrays.asList(artistIds);
        List<Artist> artists = artistRepository.findAllById(ids);

        for (Artist artist : artists) {
            artist.setIsActive(active);
        }

        artistRepository.saveAll(artists);
    }
}
